Better detection of enumerable type from an instance

paracycle opened the pull request:
<!-- (optional) Explain your change, focusing on the details of the solution. This is a great place to call out user-visible changes. -->
When detecting the type of an enumerable from an instance, `TypedEnumerable` tries to match it to a couple of well-known enumerable classes (like `Array`, `Hash`, etc). The fallback case, however, tries to match it to the class of the type that the instance is being validated against, which does not make a lot sense. 

That behaviour means, if the instance is being compared against `T::Array[Foo]`, we would try to infer the type of the enumerable instance as a kind of `T::Array`. But, if it was being compared against a `T::Set[Foo]`, then the type of the instance would be inferred to be a `T::Set`. This leads to the problem outlined in #2808.

In reality, Sorbet should not try to infer the type based on what type the instance is being compared against, but should return the type of the instance as is.

### Motivation
<!-- Why make this change? Describe the problem, not the solution. This can also be a link to an issue. -->
Fixes: #2808 

### Test plan
<!-- If you did not write tests for this change, replace the message below explaining why not. Why we should be confident this change is correct? If you changed the website, please include a screenshot of the proposed changes. -->

See included automated tests.

paracycle commented on gems/sorbet-runtime/test/types/types.rb:
```diff
@@ -442,6 +442,24 @@ def subtype_of?(other)
+        msg = type.error_message_for_obj(value)
+        expected_error = "Expected type T::Array[Integer], got Foo"
+        assert_equal(expected_error, msg)
+      ensure
+        Object.send(:remove_const, :Foo)
+      end
```
I don't like the fact that I am doing this, but I have to give a name to the class for the test to run properly, and I couldn't think of a better way of doing it. I am open to other ideas.

jez replied:
Could you define the class as a proper named class at the top of the file, sort of like a fixture?

We make a lot of helper classes in other files that way:

https://github.com/sorbet/sorbet/blob/0c301f6c2900528f9bd9a7dd869f574d4504c644/gems/sorbet-runtime/test/types/props/struct.rb#L5-L14

^ e.g., these structs are each used only in one `it` block.

paracycle replied:
Ah, makes much more sense. I'll change it the class to a fixture like those.

jez commented on gems/sorbet-runtime/test/types/types.rb:
```diff
@@ -442,6 +442,24 @@ def subtype_of?(other)
         assert_equal(0, allocs_when_invalid)
       end
+
+      it 'gives the right error when passed an unrelated enumerable' do
```
If I understand this right, it's not possible to reproduce this anymore given the recent changes to remove runtime type checking of collection types, right? e.g. #3293

jez replied:
that is, to write a test you had to call `error_message_for_obj` directly, but I don't think there's a user-visible way to trigger the behavior in #2808 anymore.

jez commented:
Your tests are failing because `Enumerable` is a module, which might have been mixed into `BasicObject` which wouldn't necessarily have a `.class` method.

Probably the easiest way around this would be to add a case for `when Object then obj.class` and otherwise do something else.

Another trick would be to `Object.instance_method(:class).bind(obj).call`.

Either one (or something else) works for me.

paracycle commented:
@jez Thanks for the review, I'll make the relevant changes. 

I believe this is still a user-visible error in the latest released version of `sorbet-runtime`, as can be seen by this repro script:
```ruby
require "bundler/inline"

gemfile do
  source 'https://rubygems.org'
  gem "sorbet-runtime", "0.5.5866"
end

require "sorbet-runtime"

class Bar
  include Enumerable

  def each
    yield "something"
  end
end

class Foo
  extend T::Sig

  sig { returns(T::Array[String]) }
  def self.foo
    Bar.new
  end
end

p Foo.foo
```
with the output:
```bash
$ ruby foo.rb
Traceback (most recent call last):
	5: from foo.rb:27:in `<main>'
	4: from /Users/ufuk/.gem/ruby/2.6.5/gems/sorbet-runtime-0.5.5866/lib/types/private/methods/_methods.rb:228:in `block in _on_method_added'
	3: from /Users/ufuk/.gem/ruby/2.6.5/gems/sorbet-runtime-0.5.5866/lib/types/private/methods/call_validation.rb:138:in `validate_call'
	2: from /Users/ufuk/.gem/ruby/2.6.5/gems/sorbet-runtime-0.5.5866/lib/types/private/methods/call_validation.rb:1125:in `report_error'
	1: from /Users/ufuk/.gem/ruby/2.6.5/gems/sorbet-runtime-0.5.5866/lib/types/configuration.rb:226:in `call_validation_error_handler'
/Users/ufuk/.gem/ruby/2.6.5/gems/sorbet-runtime-0.5.5866/lib/types/configuration.rb:219:in `call_validation_error_handler_default': Return value: Expected type T::Array[String], got T::Array[String] (TypeError)
Caller: foo.rb:27
Definition: foo.rb:22
```

jez commented:
> I believe this is still a user-visible error in the latest released version of `sorbet-runtime`, as can be seen by this repro script:

This is great, can you add this as a test case?

paracycle commented:
@jez I made the changes we discussed above and pushed this branch. Specifically:

1. I went for the `Object.instance_method(:class).bind(obj).call` to get the class name of the object
2. Left the test that I had added to assert the error message in place
3. Added a test that would exercise the signature validation failing so that we test the user-visible case as well.

Please let me know if more changes are needed. Thanks
